### PR TITLE
storage/engine: warn when committing a non-existent intent

### PR DESCRIPTION
**Pull request by petermattis**
When committing an intent, warn if a value at the associated
transaction timestamp cannot be found.

### CODE REVIEW

**Comment by petermattis** [comment]
Wrote this in an effort to track down #16266. The warnings never triggered. Not sure if this is worth committing.

File: pkg/storage/engine/mvcc.go
Comment (tamird) [inline pkg/storage/engine/mvcc.go]: Is this new code applicable in all three of these cases? I'm not exactly clear on what each of these conditions represents, but my guesses:
`!ok`: there's no value here at all
`meta.Txn == nil`: there's a value, but no intent
`!roachpb.TxnIDEqual(intent.Txn.ID, meta.Txn.ID)`: there's an intent, but it belongs to another transaction

Does `!ok` necessary mean the `mvccGetUsingIter` will return an error?

Comment (petermattis) [inline pkg/storage/engine/mvcc.go]: Your guesses are correct. `mvccGetUsingIter` should return without an error even when `!ok`. It will hit the `v == nil` case.

Comment (tamird) [inline pkg/storage/engine/mvcc.go]: Hm, ok...seems like `!ok` and v==nil are equivalent, then? picking nits, but might be clearer to do that explicitly.

**Comment by tschottdorf** [comment]
I'd be ok merging this (if things burn, it'll be good to know that this hasn't fired). Do you think there's a perf implication due to the extra read? That sits in the path of `EndTransaction` when it's not 1PC as we resolve local intents eagerly.

---

Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17580#-:-KrCX9ljrmwiPjMJ0hER:b-jgzds9)*

**Comment by petermattis** [comment]
> Do you think there's a perf implication due to the extra read? That sits in the path of EndTransaction when it's not 1PC as we resolve local intents eagerly.

Yeah, there could be. I think I could replace `mvccGetUsingIter` with `mvccGetInternal` which would remove any additional calls because the desired data would already be present. I'll do so and verify there is no performance impact before merging this. I'm primarily interested in whether I have the correct logic here.

File: pkg/storage/engine/mvcc.go
Comment (petermattis) [inline pkg/storage/engine/mvcc.go]: Done, though I don't think it really simplified anything.

**Review by tamird — APPROVED** [review]
Yep, it didn't simplify things, but the explicitness is helpful.

**Comment by bdarnell** [comment]
The logic looks right to me. I'd be OK with checking it in, but if you've already tried it on sky and it didn't fire (and we have another explanation now), I'd vote against merging this. 

---

Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17580#-:-KrNCKzqr2roEyC5T2oS:b-t6ngu8)*

**Comment by petermattis** [comment]
Yeah, I'll hold off on merging. I'm going to keep running this patch on `sky` to rule out inadvertent intent deletion as a cause of #16266. Interestingly, I haven't seen #16266 since I've been running a build which has meta2 splitting disabled.

---

Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17580#-:-KrNuYuFK4_zjeaTeMUC:b514fnh)*

**Comment by bdarnell** [comment]
> Interestingly, I haven't seen #16266 since I've been running a build which has meta2 splitting disabled.

I don't find this surprising - meta2 splits (and #2266) are a plausible cause for the "found only non-matching ranges" error in #16266 (so much so that we've retitled that issue to specifically be about meta2 splits), and that seems much more likely than a lurking bug in intent resolution.

**Comment by petermattis** [comment]
I'm not able to explain the occurrence of #16266 that is described in [this comment](https://github.com/cockroachdb/cockroach/issues/16266#issuecomment-321265199) by a meta2 split. Note that the gap in the range descriptor entries is in the middle of a meta2 range. Do you have an explanation for how a meta2 split caused that situation?

**Comment by bdarnell** [comment]
Ah, right. No, I don't have a theory for a gap in the middle of the meta range. (When did the meta2 range split? Could the userspace split that caused the gap have been concurrent with the meta2 split? I don't have a theory for how that would cause a gap here, but that's where I'd start looking)

**Comment by petermattis** [comment]
I'm planning on merging this as it did detect badness on `sky` during the meta2 debugging odyssey. The overhead should be minimal in the common case as the iterator will be positioned at the correct location making the `mvccGetInternal` call super fast (it won't actually perform a Cgo call).

**Review by tbg — APPROVED** [review]

**Pull request merged**
